File: Theory/math/permutation/iterate.py
def generate_permutations(elements):
    n = len(elements)
    c = [0] * n
    yield elements

    i = 0
    while i < n:
        if c[i] < i:
            if i % 2 == 0:
                elements[0], elements[i] = elements[i], elements[0]
            else:
                elements[c[i]], elements[i] = elements[i], elements[c[i]]
            yield elements
            c[i] += 1
            i = 0
        else:
            c[i] = 0
            i += 1

# Define a list of elements
elements = ['a', 'b', 'c']

# Generate and print permutations
for perm in generate_permutations(elements):
    print(perm)

File: Theory/math/permutation/iterate.js
function generatePermutations(arr) {
  const result = [];
  const n = arr.length;
  const c = new Array(n).fill(0);

  result.push([...arr]);

  let i = 0;
  while (i < n) {
    if (c[i] < i) {
      if (i % 2 === 0) {
        [arr[0], arr[i]] = [arr[i], arr[0]];
      } else {
        [arr[c[i]], arr[i]] = [arr[i], arr[c[i]]];
      }

      result.push([...arr]);
      c[i]++;
      i = 0;
    } else {
      c[i] = 0;
      i++;
    }
  }

  return result;
}

const inputArray = [1, 2, 3];
const permutations = generatePermutations(inputArray);
console.log(permutations);